// mjml-divider/lib/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _mjmlCore = require("mjml-core");

var _widthParser2 = _interopRequireDefault(require("mjml-core/lib/helpers/widthParser"));

var MjDivider = /*#__PURE__*/function (_BodyComponent) {
  (0, _inherits2["default"])(MjDivider, _BodyComponent);

  var _super = (0, _createSuper2["default"])(MjDivider);

  function MjDivider() {
    (0, _classCallCheck2["default"])(this, MjDivider);
    return _super.apply(this, arguments);
  }

  (0, _createClass2["default"])(MjDivider, [{
    key: "getStyles",
    value: function getStyles() {
      var _this = this;

      var p = {
        'border-top': ['style', 'width', 'color'].map(function (attr) {
          return _this.getAttribute("border-".concat(attr));
        }).join(' '),
        'font-size': '1px',
        margin: '0px auto',
        width: this.getAttribute('width')
      };
      return {
        p: p,
        outlook: (0, _objectSpread2["default"])((0, _objectSpread2["default"])({}, p), {}, {
          width: this.getOutlookWidth()
        })
      };
    }
  }, {
    key: "getOutlookWidth",
    value: function getOutlookWidth() {
      var containerWidth = this.context.containerWidth;
      var paddingSize = this.getShorthandAttrValue('padding', 'left') + this.getShorthandAttrValue('padding', 'right');
      var width = this.getAttribute('width');

      var _widthParser = (0, _widthParser2["default"])(width),
          parsedWidth = _widthParser.parsedWidth,
          unit = _widthParser.unit;

      switch (unit) {
        case '%':
          return "".concat(parseInt(containerWidth, 10) * parseInt(parsedWidth, 10) / 100 - paddingSize, "px");

        case 'px':
          return width;

        default:
          return "".concat(parseInt(containerWidth, 10) - paddingSize, "px");
      }
    }
  }, {
    key: "renderAfter",
    value: function renderAfter() {
      return "\n      <!--[if mso | IE]>\n        <table\n          ".concat(this.htmlAttributes({
        align: 'center',
        border: '0',
        cellpadding: '0',
        cellspacing: '0',
        style: 'outlook',
        role: 'presentation',
        width: this.getOutlookWidth()
      }), "\n        >\n          <tr>\n            <td style=\"height:0;line-height:0;\">\n              &nbsp;\n            </td>\n          </tr>\n        </table>\n      <![endif]-->\n    ");
    }
  }, {
    key: "render",
    value: function render() {
      return "\n      <p\n        ".concat(this.htmlAttributes({
        style: 'p'
      }), "\n      >\n      </p>\n      ").concat(this.renderAfter(), "\n    ");
    }
  }]);
  return MjDivider;
}(_mjmlCore.BodyComponent);

exports["default"] = MjDivider;
(0, _defineProperty2["default"])(MjDivider, "tagOmission", true);
(0, _defineProperty2["default"])(MjDivider, "allowedAttributes", {
  'border-color': 'color',
  'border-style': 'string',
  'border-width': 'unit(px)',
  'container-background-color': 'color',
  padding: 'unit(px,%){1,4}',
  'padding-bottom': 'unit(px,%)',
  'padding-left': 'unit(px,%)',
  'padding-right': 'unit(px,%)',
  'padding-top': 'unit(px,%)',
  width: 'unit(px,%)'
});
(0, _defineProperty2["default"])(MjDivider, "defaultAttributes", {
  'border-color': '#000000',
  'border-style': 'solid',
  'border-width': '4px',
  padding: '10px 25px',
  width: '100%'
});
module.exports = exports.default;
